expression/builtin.go

```diff
@@ -48,18 +60,38 @@ func newBaseBuiltinFunc(args []Expression, ctx context.Context) baseBuiltinFunc
 	}
 }
 
-// newBaseBuiltinFuncWithTp will be renamed to newBaseBuiltinFunc and replaces the older one later.
-// We'll move the type infer work to expression writer,
-// thus the `tp` attribute is needed for `baseBuiltinFunc`
-// to obtain the FieldType of a ScalarFunction.
-func newBaseBuiltinFuncWithTp(args []Expression, tp *types.FieldType, ctx context.Context) baseBuiltinFunc {
+// newBaseBuiltinFuncWithTp create a built-in function signature with specified types of arguments and return type.
+// argTps indicates the types of the args, retType indicates the return type of the built-in function.
+// Every built-in function needs determined argTps and retType when we create it.
+func newBaseBuiltinFuncWithTp(args []Expression, retType *types.FieldType, ctx context.Context, argTps ...argTp) (bf baseBuiltinFunc, err error) {
+	if len(args) != len(argTps) {
+		panic("unexpected length of args and argTps")
```

Comment:
len(args) **_must_** equals to len(argTps),
and `argTps` is set by the coder,
so here I use panic to give warning.